File: src/services/api-client.ts
/**
 * Cliente HTTP configurado para comunicación con la API Evora.
 * Incluye interceptors para autenticación, reintentos y manejo de errores.
 */
import axios, { AxiosError, InternalAxiosRequestConfig } from 'axios';
import { getToken, clearAll } from '@utils/secure-storage';

const API_BASE_URL = __DEV__
  ? 'http://localhost:8003/api/v2/mobile'
  : 'https://e-plataforma.com/api/v2/mobile';

const TIMEOUT_MS = 15000; // 15 segundos
const MAX_RETRIES = 2;
const RETRY_DELAY_MS = 2000;

// Callback para logout forzado (registrado por el store al iniciar)
let onUnauthorized: (() => void) | null = null;
export function setOnUnauthorized(cb: () => void) {
  onUnauthorized = cb;
}

export const apiClient = axios.create({
  baseURL: API_BASE_URL,
  timeout: TIMEOUT_MS,
  headers: {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
  },
});

// Interceptor: Añadir token Bearer a todas las peticiones
apiClient.interceptors.request.use(
  async (config: InternalAxiosRequestConfig) => {
    const token = await getToken();
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    return config;
  },
  (error) => Promise.reject(error)
);

// Interceptor: Manejo de errores y reintentos
apiClient.interceptors.response.use(
  (response) => response,
  async (error: AxiosError) => {
    const config = error.config as InternalAxiosRequestConfig & { _retryCount?: number };

    // 401: Sesión expirada → limpiar todo y forzar logout
    if (error.response?.status === 401) {
      await clearAll();
      if (onUnauthorized) onUnauthorized();
      return Promise.reject(error);
    }

    // 429: Rate limit → esperar y reintentar
    if (error.response?.status === 429) {
      const retryAfter = Number(error.response.headers['retry-after']) || 5;
      await delay(retryAfter * 1000);
      return apiClient(config);
    }

    // 5xx: Error de servidor → reintentar
    if (error.response && error.response.status >= 500) {
      config._retryCount = (config._retryCount || 0) + 1;
      if (config._retryCount <= MAX_RETRIES) {
        await delay(RETRY_DELAY_MS);
        return apiClient(config);
      }
    }

    return Promise.reject(error);
  }
);

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export default apiClient;

File: lib/core/api_client.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import 'secure_storage.dart';

/// URL base según entorno.
const String _devBaseUrl = 'http://10.0.2.2:8003/api/v2/mobile';
const String _prodBaseUrl = 'https://e-plataforma.com/api/v2/mobile';

String get apiBaseUrl => kDebugMode ? _devBaseUrl : _prodBaseUrl;

const int _timeoutMs = 15000;
const int _maxRetries = 2;
const int _retryDelayMs = 2000;

/// Callback para logout forzado (registrado por el provider de auth).
typedef OnUnauthorized = void Function();
OnUnauthorized? _onUnauthorized;

void setOnUnauthorized(OnUnauthorized cb) {
  _onUnauthorized = cb;
}

/// Cliente HTTP singleton configurado con interceptors.
final Dio apiClient = _createClient();

Dio _createClient() {
  final dio = Dio(BaseOptions(
    baseUrl: apiBaseUrl,
    connectTimeout: const Duration(milliseconds: _timeoutMs),
    receiveTimeout: const Duration(milliseconds: _timeoutMs),
    headers: {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    },
  ));

  // Interceptor: Bearer token
  dio.interceptors.add(InterceptorsWrapper(
    onRequest: (options, handler) async {
      final token = await SecureStorage.getToken();
      if (token != null) {
        options.headers['Authorization'] = 'Bearer $token';
      }
      handler.next(options);
    },
    onError: (error, handler) async {
      final response = error.response;
      final requestPath = error.requestOptions.path;

      // 401: Sesión expirada → limpiar y forzar logout
      if (response?.statusCode == 401) {
        final isAuthRoute = requestPath.contains('/auth/login') ||
            requestPath.contains('/auth/refresh');

        if (!isAuthRoute) {
          await SecureStorage.clearAll();
          _onUnauthorized?.call();
        }
        return handler.next(error);
      }

      // 429: Rate limit → esperar y reintentar
      if (response?.statusCode == 429) {
        final retryAfter =
            int.tryParse(response?.headers.value('retry-after') ?? '') ?? 5;
        await Future.delayed(Duration(seconds: retryAfter));
        try {
          final result = await dio.fetch(error.requestOptions);
          return handler.resolve(result);
        } catch (e) {
          return handler.next(error);
        }
      }

      // 5xx: Error de servidor → reintentar
      if (response != null && response.statusCode! >= 500) {
        final retryCount =
            error.requestOptions.extra['_retryCount'] as int? ?? 0;
        if (retryCount < _maxRetries) {
          error.requestOptions.extra['_retryCount'] = retryCount + 1;
          await Future.delayed(const Duration(milliseconds: _retryDelayMs));
          try {
            final result = await dio.fetch(error.requestOptions);
            return handler.resolve(result);
          } catch (e) {
            return handler.next(error);
          }
        }
      }

      handler.next(error);
    },
  ));

  return dio;
}
